Implementation of append and heredoc redirectrions

Skoteini-42 opened the pull request:
Key features :
-Heredoc (<<)
Reads input until delimiter into temporary files (/tmp/minishell_heredoc_*).
Automatic cleanup after command execution.

-Append Redirection (>>)
Opens files with O_APPEND flag to preserve existing content.
Handles multiple append/truncate redirections (last one wins).

Code changes :
-Added heredoc_del, append and heredoc_tmpfile to t_cmd struct.
-Added heredoc.c inside the parser folder which contains the following functions :
generate_unique_file, read_heredoc, process_heredoc, process_all_heredocs, cleanup_heredoc
-Updated setup_redirections function to open the outfile either in append mode (if there is an append redirection) or in trunc mode.
-Updated execute_cmd_lst function in order to process heredoc in case there is one in the command, before executing the rest of the commands.
-Updated main function in order to cleanup any existing temp files from heredoc.

Other stuff :
-Created a separate folder /parser/handlers/ and stored inside the relevant functions that parser handlers.
-Renamed cmd_list.c into inti_cmd_list.c and alocated it inside the parser folder. As a sidenote, maybe we can have a separate folder called "init" or similar to store all the init functions there.

Please feel free to do any tests and let me know, I might have missed some deltails.

Note : I noticed that if we run for eg cat << EOF | sleep 20 and we terminate the process before it properly ends using CTRL + C, the tmp heredoc file will remain inside /tmp/. That's not an issue for the functionality of the program, but if we want to be clean, prossibly this needs to be handled during the signals implementation.

Skoteini-42 committed:
committing changes before switching branch

Skoteini-42 committed:
Merge remote-tracking branch 'origin/main' into app_heredoc

Skoteini-42 committed:
updating with crrent main

Skoteini-42 committed:
added app, heredoc handlers

Skoteini-42 committed:
added app, heredoc handlers

Skoteini-42 committed:
Merge branch 'main' into app_heredoc

Skoteini-42 committed:
added herdoc porcess

Skoteini-42 committed:
adding latest changes with append

Skoteini-42 committed:
resolved confilcts

Skoteini-42 committed:
added hereredoc leanup function

Skoteini-42 committed:
Merge branch 'main' into app_heredoc

Skoteini-42 committed:
fixed norminette

Skoteini-42 committed:
fixed heredoc tmp file buffer issue

Skoteini-42 committed:
erased wrong comment made by me on the execute_cmd documentation

edouardproust commented on src/parser/cmd_utils.c:
```diff
@@ -11,8 +23,7 @@ t_cmd	*cmd_new(t_cmd *prev_cmd)
+	init_cmd_redir(cmd);
 	cmd->pipe = malloc(sizeof(int) * 2);
 	if (!cmd->pipe)
 		return (free(cmd), NULL);
```
You use several times free directly in your code: it it much safer to use a wrapper like ft_free_ptrs to prevent any issues in case of call later on in the code, when we do updates etc. 

=> I don't note this below each use of free(), you can quickly replace all the occurences of free() by ft_free_ptrs in this PR

It is true that the signature of the function is a bit complex. We could add a function like that to libft for when only one pointer is to be freed (it can be the subject of a dedicated branch later):
```
ft_free(void **ptr)
{
   ft_free_ptrs(1, ptr);
}
```

Skoteini-42 committed:
fixed fd redir

Skoteini-42 committed:
resolved leaks, fixed del issue, fixed redirections

Skoteini-42 commented:
Fixed :

Lekas --->Added inside free_cmd_list to also free the cur_cmd->heredoc_del, and cur_cmd->heredoc_tmpfile.
Added free(num_str) inside generate_unique_file to resolve leaks cause by ft_itoa

Delimiter issue --->Added ft_strcmp inside libft and changed the relevant line inside read_heredoc to call this instead ft_strncmp.

Handles multiple append/truncate redirections ---> Added (*cur_cmd)->append = 0; inside handle_redir_out that I had forgotten. Also added ft_close(&cmd->fdout); inside the conditions in setup_redirections to make sure previous fds are closed if any.

edouardproust reviewed (approved):
Ok! But i do a branch just after to replace de `free` without wrapper
